lib/internal/policy/manifest.js

```diff
@@ -0,0 +1,74 @@
+'use strict';
+//
+// {
+//   "resources": {
+//      "url": {
+//        "integrity": "sri"
+//      }
+//   }
+// }
+//
+const {
+  ERR_MANIFEST_ASSERT_INTEGRITY,
+  ERR_MANIFEST_PARSE_INTEGRITY,
+} = require('internal/errors').codes;
+const SRI = require('internal/policy/sri');
+const { SafeWeakMap } = require('internal/safe_globals');
+const crypto = require('crypto');
+const { Buffer } = require('buffer');
+const { createHash, timingSafeEqual } = crypto;
+const Hash$update = Function.call.bind(crypto.Hash.prototype.update);
+const Hash$digest = Function.call.bind(crypto.Hash.prototype.digest);
+const Buffer$toString = Function.call.bind(Buffer.prototype.toString);
+const entries = Object.entries;
+const INTEGRITIES = new SafeWeakMap();
+class Manifest {
+  constructor(obj) {
+    const integrities = {
+      __proto__: null,
+    };
+    const manifestEntries = entries(obj.resources);
+    for (var i = 0; i < manifestEntries.length; i++) {
+      const { 0: url, 1: { integrity } } = manifestEntries[i];
+      if (integrity != null) {
+        try {
+          const sri = SRI.parse(integrity);
+          integrities[url] = Object.freeze(sri);
+        } catch (e) {
+          throw new ERR_MANIFEST_PARSE_INTEGRITY(url, integrity);
+        }
+      }
+    }
+    Object.freeze(integrities);
+    INTEGRITIES.set(this, integrities);
+    Object.freeze(this);
+  }
+  assertIntegrity(url, content) {
+    const integrities = INTEGRITIES.get(this);
+    const realIntegrities = new Map();
+    if (integrities && url in integrities) {
+      const integrityEntries = integrities[url];
+      // avoid clobbered Symbol.iterator
+      for (var i = 0; i < integrityEntries.length; i++) {
+        const {
+          algorithm,
+          value: expected
+        } = integrityEntries[i];
+        const hash = createHash(algorithm);
+        Hash$update(hash, content);
+        const digest = Hash$digest(hash);
+        if (digest.length === expected.length &&
+          timingSafeEqual(digest, expected)) {
+          return true;
+        }
+        realIntegrities.set(algorithm, Buffer$toString(digest, 'base64'));
+      }
+    }
+    throw new ERR_MANIFEST_ASSERT_INTEGRITY(url, realIntegrities);
+  }
+}
+Object.setPrototypeOf(Manifest, null);
+Object.setPrototypeOf(Manifest.prototype, null);
+Object.freeze(Manifest);
+Object.freeze(Manifest.prototype);
+module.exports = Object.freeze({ Manifest });
```

Comment:
Can you add a comment on why you are freezing all this lot? I guess it's for security reasons.